### ISSUE TITLE
tmux crashes or hangs because of script used in status line

### ISSUE THREAD

**Issue by lacygoill** [tmux/tmux#1854]
### Issue description

For the last few months, tmux has been hanging or crashed at random times, and I don't know exactly why.

I have these lines in my tmux.conf:

    set -g status-interval 5
    set -g status-right '#[fg=colour235] M #[fg=colour15,bold]#(get-mem-used.sh) #[fg=colour235,nobold] C #[fg=colour15,bold]#(get-cpu-load.sh) '

The purpose of the `get-mem-used.sh` and `get-cpu-load.sh` scripts is to get a unique number representing the memory consumption and another one for the cpu load.
*The code is old, and one day I should probably use [this](https://github.com/tmux/tmux/wiki/FAQ#what-is-the-best-way-to-display-the-load-average-why-no-l) instead. But for now, I prefer my scripts.*

Each time the issue happens, a few seconds/minutes before, in the status line, the memory and cpu numbers are replaced with a message looking like this:

    <script not ready>

For some time, I thought that it was due to the `$ lscpu` shell utility, because I always had a message from the OS telling me that it crashed. But I'm not sure anymore, because I think it has also happened in the reverse order (tmux crashes, *then* lscpu crashes). Now I don't have the error message from the OS anymore, but it could be because I ticked some box to tell it to stop warning me about such crashes.

In case it helps, here's the code of `get-mem-used.sh`:

    #!/bin/bash
    used=$(free | grep Mem | awk '{ print $3 }')
    total=$(free | grep Mem | awk '{ print $2 }')
    percentage=$(bc <<< "100*${used}/${total}")
    printf -- '%s' "${percentage}"

And here's `get-cpu-load.sh`:

    #!/bin/bash
    ncore=$(lscpu | grep -i 'cpu(s)' | head -n1 | awk '{ print $2 }')
    avg1=$(cut -d' ' -f1 /proc/loadavg)
    avg1=$(bc <<< "100*${avg1}/${ncore}")
    printf -- '%s' "${avg1}"

I don't think they contain anything special.
Although, I'm using a locally compiled version of gawk, and bc. *Now that I think about it, I should make sure I'm using stable/release versions.*

Today, it has happened again; tmux didn't crash, but it didn't respond to any keypress.
When I saw the `<script not ready>` in the status line, to get a log of the tmux server, I ran:

    $ kill -s USR2 <PID>

To get a trace, I also ran:

    $ strace -ttt -ostrace.out -p <PID>

Then, after tmux started hanging, I killed `$ strace` and ran `$ gdb` to get a backtrace.
Finally, I killed the tmux server, since there was nothing more I could think of doing.

I can't reproduce at all, and it may take several weeks before the issue reappears.
It's unlikely I can provide more information before a long time, so feel free to close the issue if the current report doesn't help finding a solution.

Also, it could be that tmux is not responsible of the issue; maybe `$ lscpu`, `$ gawk` or `$ bc` is crashing and taking down `$ tmux` with it. I don't know whether it's possible.

### Required information

 - `$ tmux -V`: tmux next-3.1
 - `$ uname -sp`: Linux x86_64
 - terminal: st
 - `$TERM` outside tmux: st-256color
 - `$TERM` inside tmux: tmux-256color
 - log from `$ kill -s USR2 PID`:
[tmux-server-2535.log](https://github.com/tmux/tmux/files/3414504/tmux-server-2535.log)
 - trace from `$ strace -ttt -ostrace.out -p PID`:
[strace.txt](https://github.com/tmux/tmux/files/3414507/strace.txt)
 - backtrace from `$ gdb -q /usr/local/bin/tmux PID`, `set logging on`, `bt full`:
[backtrace.txt](https://github.com/tmux/tmux/files/3414506/backtrace.txt)

**Comment by lacygoill** [comment]
I've reinstalled `$ bc` and `$ gawk` to make sure I was using stable versions.
I'm also going to rewrite the scripts to invoke less shell utilities (`$ grep` is probably useless), taking inspiration from the snippet in the FAQ.

Maybe this will fix the issue.

Also, I've just noticed that the backtrace contains a lot of `<optimized out>`, this is because I compiled with `-O2` instead of `-O0`, sorry.
I've recompiled with `-O0`, and edited my installation script so that it edits `Makefile.am` automatically.

I'll reopen the issue if tmux hangs or crashes again in the future. If that happens, the backtrace will contain more useful information.

**Comment by lacygoill** [comment]
Btw, I have this alias definition in my tmux.conf:

    if '[ "$TERM" != #{default-terminal} ]' {
        set -as command-alias log={if '[ ! $(pidof strace) ]' \
        { display 'Started tmux server logging' ; run 'kill -s USR2 #{pid} ; strace -ttt -ostrace.txt -p #{pid}' } \
        { display 'Stopped tmux server logging' ; run 'kill -s USR2 #{pid} ; killall strace' }
    }}

I've created it for the purpose of debugging this issue.
I've tested it and it works.
But yesterday, right before tmux started to hang, when I ran it with `$ tmux log`, it didn't work.
The file `strace.txt` was not created and `$ strace` was not running.

So now I'm going to test this zsh function instead:

    tmux_log() {
      emulate -L zsh
      if pidof strace >/dev/null 2>&1; then
        tmux display 'Stopped tmux server logging' \; \
             run 'kill -s USR2 #{pid} ; killall strace'
      else
        tmux display 'Started tmux server logging' \; \
             run 'kill -s USR2 #{pid} ; strace -ttt -ostrace.txt -p #{pid}'
      fi
    }

**Comment by lacygoill** [comment]
as inefficient to re-run `$ lscpu` every time tmux updates the status line, since I only run it to get the number of cpu cores.

So now, I cache this number in a file `~/.ncore`:

    if '[ "$TERM" != "#{default-terminal}" -a ! -s "${HOME}/.ncore" ]' \
        { run "lscpu | awk '/^CPU\\(s\\):\\s*[0-9]/ { print $2 }' >\"${HOME}/.ncore\"" }

Also, to reduce the amount of time and code tmux has to run, I've deleted the scripts, and moved the shell commands directly inside the value of `'status-right'`:

    cpu='awk '\''BEGIN { \
        getline load <"/proc/loadavg"; \
        getline ncore <(ENVIRON["HOME"]"/.ncore"); \
        printf("%%d", 100 * load / ncore) }'\'''
    mem='free | awk '\''/Mem/ { total = $2; used = $3 }; END { printf("%%d", 100 * used / total) }'\'''
    set -g  status-right '#[fg=colour235]M #[fg=colour15,bold]'
    set -ga status-right "#($mem) "
    set -ga status-right '#[fg=colour235,nobold]C #[fg=colour15,bold]'
    set -ga status-right "#($cpu) "
    setenv -gu cpu
    setenv -gu mem

This way, tmux doesn't have to invoke sh to interpret the script.

It seems to have almost cut in half the amount of time it takes to run the shell command:

    $ cat <<'EOF' >/tmp/sh.sh
    awk 'BEGIN { getline load <"/proc/loadavg"; getline ncore <(ENVIRON["HOME"]"/.ncore"); printf("%d", 100 * load / ncore) }'
    EOF
    $ chmod +x /tmp/sh.sh

    $ time zsh -c 'repeat 100 /tmp/sh.sh'
    ... 0,420 total

    $ time zsh -c 'repeat 100 awk '\''BEGIN { getline load <"/proc/loadavg"; getline ncore <(ENVIRON["HOME"]"/.ncore"); printf("%d", 100 * load / ncore) }'\'''
    ... 0,193 total

**Comment by lacygoill** [comment]
It happened again today. When the message appeared in the status line, I made the mistake of pressing my key binding to reload the config which always make tmux hang in such a case.

So, all the refactoring of `'status-right'` didn't fix the issue.
I'm not using a script anymore, and the only shell utilities invoked by tmux are `$ free` and `$ awk`.
`$ free` comes from the default package `procps`, and `$ awk` from a locally compiled version of gawk 5.0.1 (stable release downloaded from [here](https://ftp.gnu.org/gnu/gawk/)):

    $ free -V
    free from procps-ng 3.3.10

    $ awk -V | head -n1
    GNU Awk 5.0.1, API: 2.0 (GNU MPFR 3.1.4, GNU MP 6.1.0)

I should not have resourced my tmux.conf; instead I should have waited for the crash to get a core and a backtrace. Next time, I will try to remember to avoid doing that.

I didn't have this issue a year ago, so I'm starting to think a regression has been introduced in the last months, but since the issue only occurs once in a month, it's hard to get a good estimate of those durations.

**Comment by lacygoill** [comment]
@nicm: I have the issue right now, but I don't know what to do.

The right part of the status line displays:

    M <'free | awk '/Mem/ { total = $2; used

This is the beginning of the awk code I use to get the memory consumption on my system.

So it has become useless, because I can't see the memory or cpu load.
I don't want to reload the tmux config because it would make tmux freeze, and I wouldn't get a core from which I could extract a backtrace.

I've run my debugging shell script; it has started `$ strace`, but it hasn't logged the activity of the tmux server. I've tried to run `$ tmux kill -s USR2 $(tmux display -p '#{pid}')` and `$ tmux kill -s USR2 2678` (2678 is the pid of my current tmux server), but none has created the ususal `tmux*.log` file in my home directory (from which I always start tmux).
I've run `$ sudo updatedb` and `$ locate -r 'tmux.*.log' | grep 2678` to try and find the log file elsewhere, but the last command has an empty output.

Here are the last 10000 lines of another strace file, in addition to the one included in the OP.
[strace.txt](https://github.com/tmux/tmux/files/3511734/strace.txt)

Tmux doesn't seem to crash.
What else can I do to get more information and debug this issue?

**Comment by nicm** [comment]
What does tmux info show?

**Comment by lacygoill** [comment]
Here is a file containing the output of `$ tmux info`: [tmux_info.txt](https://github.com/tmux/tmux/files/3511784/tmux_info.txt)

**Comment by nicm** [comment]
You are running the kill and strace from #() as well?

Did you try attaching gdb to the tmux server process after it hangs to get
a backtrace?

**Comment by lacygoill** [comment]
> You are running the kill and strace from #() as well?

No, I ran them from a sh script started from a zsh command-line.

> Did you try attaching gdb to the tmux server process after it hangs to get a backtrace?

I was waiting for a crash because I assumed that it would lead to more information.

I have reloaded my tmux.conf, and as usual it makes tmux hang.
I have then tried to get a backtrace:

    $ gdb /usr/local/bin/tmux 2678
    (gdb) set logging on
    (gdb) bt full
    (gdb) quit

But `$ gdb` told me that there was no stack.
The message suggested to retry as the root user, so I tried the same commands with `$ sudo`, but there was still no stack. I think the issue was because there was already a process running the ptrace(2) system call, so I killed it.

I then reran `$ gdb` and here's the backtrace: [gdb.txt](https://github.com/tmux/tmux/files/3511801/gdb.txt)

**Comment by nicm** [comment]
This trace is from after tmux is hanging? What libevent version are you
using?

**Comment by lacygoill** [comment]
> This trace is from after tmux is hanging?

Yes, I attached gdb to the tmux server which is still currently hanging.

> What libevent version are you using?

I'm not sure, but the backtrace referred to the file `/lib/x86_64-linux-gnu/libevent-2.0.so.5`. And according to `$ apt-file search`, it comes from the package `libevent-2.0-5`. Here's how `$ aptitude show libevent-2.0-5` describes the package:

    $ aptitude show libevent-2.0-5
    Package: libevent-2.0-5
    State: installed
    Automatically installed: no
    Multi-Arch: same
    Version: 2.0.21-stable-2ubuntu0.16.04.1
    Priority: optional
    Section: libs
    Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
    Architecture: amd64
    Uncompressed Size: 306 k
    Depends: libc6 (>= 2.17)
    Breaks: libevent-2.0-5:i386 (!= 2.0.21-stable-2ubuntu0.16.04.1)
    Replaces: libevent-2.0-5:i386 (< 2.0.21-stable-2ubuntu0.16.04.1)
    Description: Asynchronous event notification library
     Libevent is an asynchronous event notification library that provides a
     mechanism to execute a callback function when a specific event occurs on a file
     descriptor or after a timeout has been reached.

     It is meant to replace the asynchronous event loop found in event driven
     network servers. Currently, libevent supports /dev/poll, kqueue(2), event
     ports, select(2), poll(2) and epoll(4).

     For backward compatibility and to transition to "libevent_core only" or
     "libevent_core + libevent_extra", the libevent source package builds the
     libevent library that includes everything in libevent_core and libevent_extra.
    Homepage: http://libevent.org/

**Comment by lacygoill** [comment]
I think Ubuntu 16.04 installs libevent version 2.0.21 because of this line in the output of the last command:

    Version: 2.0.21-stable-2ubuntu0.16.04.1

It's seven years old. Does it mean that I should update libevent to fix the issue?

The latest stable release (2.1.11) is less than one month old.

**Comment by nicm** [comment]
I would try updating to libevent 2.1 and see if it fixes the problem.

**Comment by lacygoill** [comment]
I know how to update libevent: <https://github.com/tmux/tmux/issues/1840#issuecomment-513273184>

But there are too many things that I don't understand in the commands.
I will keep the habit of quitting the server and restarting it whenever the issue occurs, until I update to Ubuntu 18.04 or maybe Ubuntu 20.04 next year. On Ubuntu 18.04, libevent is only 2 years old.

Thank you for the help.

**Comment by lacygoill** [comment]
@nicm: Even though I don't understand all the commands, I've decided to compile the latest stable release of libevent (2.1.11), and re-compile tmux so that it uses this new libevent:

    wget https://github.com/libevent/libevent/releases/download/release-2.1.11-stable/libevent-2.1.11-stable.tar.gz
    tar -zxvf libevent-2.1.11-stable.tar.gz
    cd libevent-2.1.11-stable
    CFLAGS="-g -O0" ./configure --enable-shared --prefix=$HOME/libeventbuild --disable-silent-rules
    make
    make install
    cd ..
    git clone https://github.com/tmux/tmux.git
    cd tmux
    sh autogen.sh
    PKG_CONFIG_PATH=$HOME/libeventbuild/lib/pkgconfig ./configure --prefix=$HOME/tmuxbuild
    make

I've also installed this shell alias, to start tmux more easily:

    alias tmux='LD_LIBRARY_PATH=~/libeventbuild/lib ~/GitRepos/tmux/tmux'

But the issue seems to persist; the right part of my status line has just started to show:

    M <'free | awk '/Mem/ { total = $2; used

After that, resourcing `~/.tmux.conf` made tmux hang as usual.

Here is the output of `$ tmux info`: [tmux_info.txt](https://github.com/tmux/tmux/files/3541257/tmux_info.txt)
And here is a backtrace from `$ gdb`: [backtrace.txt](https://github.com/tmux/tmux/files/3541259/backtrace.txt)

**Comment by nicm** [comment]
Does it stop hanging if you kill the two PIDs at the end of tmux info?

This gdb backtrace is from the client, can you get the backtrace from the server?

**Comment by nicm** [comment]
What if you start tmux from the beginning with -vv and then reproduce, can you show me the full server log from that?

**Comment by lacygoill** [comment]
> Does it stop hanging if you kill the two PIDs at the end of tmux info?

No, but I've just noticed something (I don't know whether this is new; it could be due to the newer version of libevent but I'm not sure): if I close the terminal window (Alt-F4), then open a new one, and run `$ LD_LIBRARY_PATH=~/libeventbuild/lib ~/GitRepos/tmux/tmux attach`, then tmux stops hanging.

> This gdb backtrace is from the client, can you get the backtrace from the server?

I don't think it's the backtrace from the client, for 2 reasons:

   - I got the pid 2201 from the output of `$ tmux display -p '#{pid}'`, and according to `$ man tmux`, the format variable `#{pid}` should be replaced with the pid of the server, not the client

   - when I run `$ ps aux | grep 2201`, in the `TTY` section, `?` is written; if it was the client, something like `pts/123` would be written (not `?`)

That being said, if you need a backtrace from a client, and not from the server, I think I have one: [backtrace.txt](https://github.com/tmux/tmux/files/3541380/backtrace.txt)

**Comment by lacygoill** [comment]
> What if you start tmux from the beginning with -vv and then reproduce, can you show me the full server log from that?

I guess I could, but I can't reproduce on demand, and the logfile will probably be huge at the end of each day (I don't have much space on my ssd).

I will still try to tweak my shell alias to include the `-vv` flag, and if I can, the next time the issue appears, I'll post the compressed logfile (or the last 100000 lines if it's too big).

**Comment by lacygoill** [comment]
I had another idea. I thought I could send `SIGUSR1` to the tmux server (pid 2201), which should be equivalent to the `-v` flag. Then resource my `~/.tmux.conf` to make tmux hang.
So I ran:

    $ kill -s USR1 2201

Unfortunately, it doesn't work, the logfile is not created.
I think it should be created in the directory from which the tmux server was started (that is in my home), but I can't find it anywhere, nor with:

    $ sudo updatedb && locate -r 'tmux.*log'

I always have this other issue when the right part of the status line stops working; i.e. `SIGUSR1` seems to be ignored.

**Comment by nicm** [comment]
Oh yeah you're right its the server backtrace. Try and run tmux with -vv and reproduce and get the full log.

**Comment by lacygoill** [comment]
Now that I can make tmux stops hanging (by attaching a new client), I've been able to make more experiments.
In particular, I wanted to know which lines exactly in `~/.tmux.conf` made tmux hang.
Here's what I found:

    if '[ ! -s "${HOME}/.ncore" ]' \
        { run "lscpu | awk '/^CPU\\(s\\):\\s*[0-9]/ { print $2 }' >\"${HOME}/.ncore\"" }
    if '[ "$TERM" != "#{default-terminal}" -a -n "$DISPLAY" ]' { source "$HOME/.tmux/terminal-overrides.conf" }
    if '[ "$TERM" != "#{default-terminal}" ]' { set -ga update-environment COLORTERM }
    if '[ "$TERM" != "#{default-terminal}" ]' { source "$HOME/.tmux/plugins/run" }
    if '[ "$TERM" != "#{default-terminal}" ]' { source "$HOME/.tmux/unbind-copy-mode.conf" }

Each of these lines make tmux hang.
They all have in common the `if-shell` command, so I think that any tmux command which starts a shell process makes tmux hang once the issue appears.

And indeed, if I run a simple `$ tmux run ''`, or `$ tmux if true ''`, the tmux server does not hang, because I can still interact with it, but these particular processes keep running. I can't end them by pressing `C-c`, `C-d` or `C-\`. The only way I can end them, is by running `$ kill PID`, where PID is their pid (found with `$ ps aux | grep tmux`).

I'm not sure, but maybe it means that when I resource my tmux.conf, it's the client which is hanging and not the server, if that makes any sense.

---

I've run quite a few `if` and `run` commands, and now I have 51 jobs in the output of `$ tmux info`.
According to `$ ps aux | grep PID`, they are all zombies, because of the `Z` flag in the STAT column, and the `<defunct>` indicator in the COMMAND column:

    user     19916  0.0  0.0      0     0 ?        Z    16:04   0:00 [sh] <defunct>
                                                   ^                      ^^^^^^^^^

Their parent is the tmux server (pid 2201).

I've tried to make the tmux server reap its dead children by sending it the signal `SIGCHLD`:

    $ kill -s CHLD 2201

But it didn't work. The zombies are still present in the output of `$ ps aux`.

I've then tried to make it reap its dead children by running:

    $ gdb -p 2201
    (gdb) call waitpid(19916, 0, 0)
    (gdb) quit

*Here `19916` is the pid of a zombie.*
It worked, because the zombie is not in the output of `$ ps aux` anymore.
However, it's still in the output of `$ tmux info`; I don't understand why.

---

Note that `run-shell` (and probably `if-shell` too) do succeed in runnning a shell command. For example, if I run:

    $ tmux run 'echo test >/tmp/file'

`/tmp/file` is correctly created with the contents `test`.
The issue is that:

   - The shell process becomes a zombie.
   - The `$ tmux run` process does not terminate, even if I make the tmux server reap the shell zombie with `call waitpid(PID, 0, 0)`.
   - I can't include the output of any shell process in the status line.
     As an example, `$ uptime|awk '{split(substr($0, index($0, "load")), a, ":"); print a[2]}'` copied from [here](https://github.com/tmux/tmux/wiki/FAQ#what-is-the-best-way-to-display-the-load-average-why-no-l) doesn't work either.

---

> Try and run tmux with -vv and reproduce and get the full log.

Ok, I will try.
If you don't mind, I'll keep the issue opened until the issue reappears; it could take a few days or a few weeks, I don't know. If I can't get the log because it takes too much space, I'll close the issue.

**Comment by nicm** [comment]
if-shell and run-shell will block the client command queue where the source-file happens until the child process exits so the hanging but being able to attach a new client is consistent with tmux not realising that the process has exited.

**Comment by nicm** [comment]
You can see the same if you have a file that does 'run "sleep 10"' then do C-b : source myfile - the client won't respond until the sleep is finished.

Now the question is why tmux is not realizing these processes have exited.

Can you reproduce this with simple commands rather than complex pipelines, for example just using echo instead of lscpu and awk?

**Comment by nicm** [comment]
It's interesting that the processes are zombies but it doesn't wait() for them even if you send it SIGCHLD manually.

**Comment by lacygoill** [comment]
> if-shell  and  run-shell  will  block   the  client  command  queue  where  the
> source-file happens until the child process exits so the hanging but being able
> to attach a new  client is consistent with tmux not  realising that the process
> has exited.

I see, thank you for the explanation. So that does mean that it's not the server which is hanging, but just the client; probably because it doesn't realize the shell process has finished its task.

> Can you reproduce this with simple  commands rather than complex pipelines, for
> example just using echo instead of lscpu and awk?

I don't think I want to do it right now.
I've killed the tmux server half an hour ago, and I don't know when the issue will reappear; it could be today, or next month.
IIUC, I would need to tweak `status-right` in my tmux.conf so that it merely `$ echo` some string. I would lose the cpu load and memory consumption indicators for an indeterminate amount of time.

I'll do it if that's really necessary. But before that I'll try to produce new logs; I've just started a new server with `-vv`.

Also, it's true that I'm using a somewhat complicated `$ awk` shell pipeline, but I'm not using `$ lscpu` anymore. I've cached its output in a file, and I make `$ awk` read it.
The only utilities that I ask tmux to run are `$ awk` and `$ free`.

> It's interesting that the processes are zombies but it doesn't wait() for them
> even if you send it SIGCHLD manually.

When the issue appears, the tmux server doesn't respond to `SIGUSR1` either, so I'm starting to think that it doesn't respond to any signal at all, but I could be wrong.

**Comment by nicm** [comment]
No it is a server problem but only one client will hang.

If it isn't getting SIGCHLD it will never know the processes have exited so it will never let the client continue.

**Comment by lacygoill** [comment]
I have the issue right now.

Here are the full logs generated by the flag `-vv`:
 - [tmux-client-1828.log](https://github.com/tmux/tmux/files/3562746/tmux-client-1828.log)
 - [tmux-out-1830.log.gz](https://github.com/tmux/tmux/files/3562747/tmux-out-1830.log.gz)

The server log takes around 1.7 Gig; I've compressed it with `$ gzip`:

    $ gzip -9 -c tmux-server-1830.log > tmux-server-1830.log.gz

But the archive still takes around 170 Megs; so I don't think I can attach here directly, since the limit is 10 or 25 Megs. Instead, I uploaded it on 0x0.st: https://0x0.st/z4tD.gz

**Comment by lacygoill** [comment]
To be clear, at the time I copied the logs, tmux was not hanging, because I hadn't reloaded my tmux.conf nor run any `if-shell` or `run-shell` command. But the status line was displaying:

    M <'free | awk '/Mem/ { total = $2; used

And the output of `$ tmux info` contained a job whose pid matched a zombie process.

**Comment by lacygoill** [comment]
Let me know if you need the logs while a tmux client is hanging.
Or if you need a backtrace of the server or of the client.
I'll keep the server running a few hours; after that, I'll have to kill it when shutting down my pc.

**Comment by nicm** [comment]
I am away this week and won't look at this until I get home, sorry.

**Comment by lacygoill** [comment]
Ok, no worry. I'll save another set of newer logs, just in case the ones I posted don't contain enough information.

**Comment by nicm** [comment]
So if I pick one of the zombies, tmux is seeing an error on its output, so
~~~~
1567283301.844008 run job 0x22250e0: [ "$TERM" != "tmux-256color" -a -n "$DISPLAY" ], pid 25203
1567283301.849492 job write 0x22250e0: [ "$TERM" != "tmux-256color" -a -n "$DISPLAY" ], pid 25203, output left 0
1567283301.849517 job error 0x22250e0: [ "$TERM" != "tmux-256color" -a -n "$DISPLAY" ], pid 25203
~~~~
But it is either never getting it returned from waitpid() or is not interpreting that correctly. So it is still left around as a zombie and in the tmux job list:
~~~~
Job 2: [ "$TERM" != "tmux-256color" -a -n "$DISPLAY" ] [fd=20, pid=25203, status=0]
~~~~
I don't know why tmux isn't seeing this. Can you apply this please then reproduce and show me the server log, tmux info and ps aux output again (I don't need client logs or backtrace)?
~~~~
diff --git a/server.c b/server.c
index a5d5e37f..0ae70a51 100644
--- a/server.c
+++ b/server.c
@@ -417,7 +417,9 @@ server_child_signal(void)
 	pid_t	 pid;
 
 	for (;;) {
-		switch (pid = waitpid(WAIT_ANY, &status, WNOHANG|WUNTRACED)) {
+		pid = waitpid(WAIT_ANY, &status, WNOHANG|WUNTRACED);
+		log_debug("%s: waitpid %ld %#x", __func__, (long)pid, status);
+		switch (pid) {
 		case -1:
 			if (errno == ECHILD)
 				return;

**Comment by lacygoill** [comment]
I have applied the patch against [2e90841](https://github.com/tmux/tmux/commit/2e90841f2ed4fbe0bb13681c9a90e355fe98ab1d), recompiled tmux, and restarted the server using my shell alias.

Since you don't need the client log, I guess you don't need tmux output either, so I replaced `-vv` with `-v` in my shell alias:

    alias tmux='LD_LIBRARY_PATH=~/libeventbuild/lib ~/GitRepos/tmux/tmux -v'
                                                                          ^

I will post the necessary information as soon as the tmux server has zombie processes.

**Comment by lacygoill** [comment]
I had the issue an hour ago.

Here is an archive containing the server log, as well as the output of `$ tmux info` and `$ ps aux`: https://0x0.st/zy4r.gz

**Comment by nicm** [comment]
Hmm it definitely looks like tmux is not getting or not handling the SIGCHLD.

Next time it happens can you get the PID of the tmux server and show me the contents of /proc/PID/status? (If you aren't sure which is the tmux server then show me it for all the tmux PIDs).

**Comment by lacygoill** [comment]
Ok, next time I'll also include a `proc_status.txt` file created with:

    $ cat /proc/$(tmux display -p '#{pid}')/status >/tmp/proc_status.txt

**Comment by lacygoill** [comment]
I have the issue right now.

Here is a link to a copy of [`/proc/790/status` right now](https://0x0.st/zyEu.txt).
*790 is the pid of the tmux server.*

Here is a link to a copy of [`/proc/790/status` as it was initially when I started the server](https://0x0.st/zyEQ.txt).

I have a [bash script](https://0x0.st/zyE1.bin) to translate the bitmasks `SigBlk`, `SigIgn`, `SigCgt` into a human-readable form.

Here are the signals which the tmux server [blocked/ignored/caught when I started it](https://0x0.st/zyE_.txt).
And here are the signals which the tmux server [blocks/ignores/catches right now](https://0x0.st/zyEp.txt).

As you can see, it still ignores and catches the same signals, but for some reason, now it blocks the following signals:

   - `ABRT`
   - `ALRM`
   - `BUS`
   - `CHLD`
   - `CONT`
   - `PIPE`
   - `STKFLT`
   - `TSTP`
   - `URG`
   - `USR1`
   - `USR2`
   - `VTALRM`

The list includes `CHLD` (which probably explains the zombies) and `USR2` (which probably explains why I can't toggle the server logging).

**Comment by lacygoill** [comment]
There are two other fields which may be interesting, `SigQ` and `ShdPnd`.

Initially:

    SigQ:	1/14291
    ShdPnd:	0000000000000000

When I had the issue:

    SigQ:	13/14291
    ShdPnd:	0000000000010000

According to `man 5 proc`, `SigQ` is the number of currently queued signals; so when I had the issue there were 13 signals in a queue instead of 1. Not sure what can make the queue of signals increase.

And `ShdPnd` is the number of signals pending for the process as a whole.
If `0000000000010000` is a hex number, it's equivalent to `65536` in decimal.
If that means that there were 65536 signals which the tmux server had to process when the issue occurred, that seems to be an abnormally large number.

**Comment by krader1961** [comment]
`ShdPnd` is not a count. It is a bitmask of the signals pending delivery. The value 0x10000 means `SIGCHLD` is pending delivery. Which isn't surprising given the nature of the problem and the fact that  it is blocked.

**Comment by lacygoill** [comment]
It happened again: [/proc/9046/status](https://0x0.st/zyI1.txt).

Here is the list of signals which are blocked this time:

   - `ALRM`
   - `BUS`
   - `CHLD`
   - `SEGV`
   - `STKFLT`
   - `TERM`
   - `TRAP`
   - `TTOU`
   - `USR1`
   - `USR2`
   - `XFSZ`

**Comment by krader1961** [comment]
You'll want to focus on the four places in the code that use `SIG_BLOCK`. I quickly skimmed the code and didn't see any situations that don't correctly unblock `SIGCHLD`.  But given the behavior you're seeing, @lacygoill, there must be one such mismatch between blocking and unblocking the signal.

**Comment by nicm** [comment]
I think there is a problem with splitw -I. Try this please: [x.diff.txt](https://github.com/tmux/tmux/files/3619654/x.diff.txt)

**Comment by lacygoill** [comment]
I have applied the patch. If nothing happens in the next few weeks, I'll report it back here.

**Comment by lacygoill** [comment]
When you wrote there was a problem with `splitw -I`, I read it as `split -l`, and I didn't understand how the `-l` flag could cause the current issue.

Then I saw the message of 697f9383550a8bb60ad09c22a6c208a0cf151e4e which refers to an empty pane.
This made me remember something. A few months ago, I discovered the concept of empty pane and decided to use it in one of my [Vim plugins](https://github.com/lacygoill/vim-logevents/blob/90eb3a7a6bb8d17460a97bc6b8142a7e5f2a9b39/autoload/logevents.vim#L177) to log all the Vim events fired during an interactive session (in the past, I had to create a temporary file instead).

It's possible that my issue started in the same time frame, and that it was caused by the empty pane created by the Vim plugin.
To check this hypothesis, I've checked out the previous commit, started Vim, and used my plugin to log some Vim events. After quitting Vim, I've checked how the tmux process reacts to signals; and indeed, it blocks several signals. Sometimes, those signals include `CHLD`, but not always.

Anyway, I found a way to reproduce the issue frequently.
So, I've decided to check how tmux behaves after 697f9383550a8bb60ad09c22a6c208a0cf151e4e, and this time, tmux doesn't block signals after using my Vim plugin.

From the start, I've conflated the consequence of the issue with its cause.
Tmux wrongly blocking `CHLD` caused the error message in the status line.
But since I only saw the error message about the script (not the signals disposition, nor the zombie processes), and my bash script was so badly written, I was sure it was the script which was the cause.

Sorry for the confusion, and thank you very much for the fix.
